pkg/kademlia/queue.go

```diff
@@ -11,28 +11,28 @@ import (
 	"storj.io/storj/pkg/pb"
 )
 
-//XorQueue is a priority queue where the priority is key XOR distance
+// XorQueue is a priority queue where the priority is key XOR distance
 type XorQueue struct {
 	maxLen int
 	items  items
 }
```

Comment:
```
type XorQueue struct {
	maxLen int

	mu    sync.Mutex
	added map[string]int
	items items
}

...

func (x *XorQueue) Insert(target dht.NodeID, nodes []*pb.Node) {
    x.mu.Lock()
    defer x.mu.Unlock()

    unique := nodes[:0]
    for _, node := range nodes {
        if _, added := x.added[node.GetId()]; !added {
            x.added[node.GetId()]++
            unique = append(unique, node)
        }
    }

    x.insert(target, nodes)
}

func (x *XorQueue) Reinsert(target dht.NodeID, node *pb.Node, limit int) {
    x.mu.Lock()
    defer x.mu.Unlock()

    if x.added[id] >= limit {
        return
    }
    x.added[id]++
    
    x.insert(target, []*pb.Node{node})
}

// insert must hold lock while adding
func (x *XorQueue) insert(target dht.NodeID, nodes []*pb.Node) {
    ... // same as Insert
}
```

Then you won't need the contacted map on the XorQueue, it will ensure that nodes don't get double added and that they are accessed at most limit times.